src/Kestrel.Core/Internal/HttpConnection.cs

```diff
@@ -143,15 +144,30 @@ public Task StartRequestProcessing<TContext>(IHttpApplication<TContext> applicat
                     adaptedPipelineTask = adaptedPipeline.RunAsync(stream);
                 }
 
-                if (_http1Connection.ConnectionFeatures.Get<ITlsApplicationProtocolFeature>()?.ApplicationProtocol == "h2" &&
-                    Interlocked.CompareExchange(ref _http2ConnectionState, Http2ConnectionStarted, Http2ConnectionNotStarted) == Http2ConnectionNotStarted)
+                var hasTls = _context.ConnectionFeatures.Get<ITlsConnectionFeature>() != null;
+                var applicationProtocol = _context.ConnectionFeatures.Get<ITlsApplicationProtocolFeature>()?.ApplicationProtocol;
+                var http1Enabled = (_context.Protocols & EndPointProtocols.Http1) == EndPointProtocols.Http1;
+                var http2Enabled = (_context.Protocols & EndPointProtocols.Http2) == EndPointProtocols.Http2;
+
+                if (!hasTls && http1Enabled && http2Enabled)
+                {
+                    throw new InvalidOperationException("HTTP/1.x and HTTP/2 cannot both be served on the same cleartext endpoint.");
+                }
+
+                var useHttp2 = http2Enabled && (!hasTls || applicationProtocol == "h2");
```

Comment:
Should we use byte arrays for these values? As discussed with @Tratcher yesterday, ALPN just uses byte strings for protocols rather than specifying strings and encodings and such.

Reply:
I'm not immediately concerned with this as long as it doesn't leak into public APIs. We'll have to refactor when we get the real ALPN support anyways.

Reply:
`ITlsApplicationProtocolFeature` is (and will be released someday as) a public API. It's a generic ALPN feature for TLS connections.

Reply:
There's a new type coming that we can use for the datatype here so it doesn't have to be raw bytes.
https://github.com/dotnet/corefx/issues/23177 or we can go down one layer to `ReadOnlyMemory<byte>`.

Reply:
I think we'd need to go to `ReadOnlyMemory<byte>`. We'd want this interface to be `netstandard` even though the implementation will be `netcoreapp2.x` until the SslStream features reach `netstandard`. So unless the exchange type is moved to a different `netstandard`-compatible assembly, I think we'll have to bypass it.